src/main/java/org/cryptacular/util/CertUtil.java

```diff
@@ -322,7 +322,7 @@ public static X509Certificate[] readCertificateChain(final InputStream in) throw
     try {
       final CertificateFactory factory = CertificateFactory.getInstance("X.509");
       final Collection<? extends Certificate> certs = factory.generateCertificates(in);
-      return certs.toArray(new X509Certificate[certs.size()]);
```

Comment:
I recall that the creating a zero-length array is the recommended way to do this (pre-Java 11), but I sincerely can't imagine that it would be more efficient than simply creating the array of the correct size that is known with certainty at the outset.

Reply:
The JVM can optimize array creation, but only if you don't create one yourself. (At least that's my understanding) And yes, it is counter intuitive.